openmc/lattice.py

```diff
@@ -0,0 +1,1901 @@
+
+                    # Handle the degenerate center ring separately.
+                    u_id = universe_ids[z, a, x]
+                    uarray[-1].append([universes[u_id]])
+            else:
+                # Build array of Universe pointers for the Lattice.  Note that
+                # we need to convert between the HDF5's square array of
+                # (alpha, y, z) to the Python API's format of a ragged nested
+                # list of (z, ring, theta).
+                uarray = []
+                for z in range(n_axial):
+                    # Add a list for this axial level.
+                    uarray.append([])
+                    a = 2*n_rings - 2
+                    y = n_rings - 1
+                    for r in range(n_rings - 1, 0, -1):
+                        # Add a list for this ring.
+                        uarray[-1].append([])
+
+                        # Climb down the bottom-right.
+                        for i in range(r):
+                            uarray[-1][-1].append(universe_ids[z, y, a])
+                            y -= 1
+
+                        # Climb across the bottom.
+                        for i in range(r):
+                            uarray[-1][-1].append(universe_ids[z, y, a])
+                            a -= 1
+
+                        # Climb up the bottom-left.
+                        for i in range(r):
+                            uarray[-1][-1].append(universe_ids[z, y, a])
+                            a -= 1
+                            y +=1
+
+                        # Climb up the top-left.
+                        for i in range(r):
+                            uarray[-1][-1].append(universe_ids[z, y, a])
+                            y += 1
+
+                        # Climb across the top.
+                        for i in range(r):
+                            uarray[-1][-1].append(universe_ids[z, y, a])
+                            a += 1
+
+                        # Climb down the top-right.
+                        for i in range(r):
+                            uarray[-1][-1].append(universe_ids[z, y, a])
+                            a += 1
+                            y -= 1
+
+                        # Move down to the next ring.
+                        a -= 1
+
+                        # Convert the ids into Universe objects.
+                        uarray[-1][-1] = [universes[u_id]
+                                          for u_id in uarray[-1][-1]]
+
+                    # Handle the degenerate center ring separately.
+                    u_id = universe_ids[z, y, a]
+                    uarray[-1].append([universes[u_id]])
+
+            # Add the universes to the lattice.
+            if len(pitch) == 2:
+                # Lattice is 3D
+                lattice.universes = uarray
+            else:
+                # Lattice is 2D; extract the only axial level
+                lattice.universes = uarray[0]
+
+        return lattice
```

Comment:
If an invalid lattice type (neither of `rectangular` or `hexagonal`) is provided this will throw an "UnboundLocalError" as `lattice` would be referenced before assignment.